File: src/main/java/baekjoon/dp/카드_구매하기2.java
package baekjoon.dp;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

/**
 * https://www.acmicpc.net/problem/16194
 *
 */
public class 카드_구매하기2 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        카드_구매하기2 fn = new 카드_구매하기2(Arrays.stream(br.readLine().split(" ", n))
                .mapToInt(Integer::parseInt)
                .toArray());
        System.out.println(fn.bottomUp(n));
    }

    private int[] prices;

    public 카드_구매하기2(int... n) {
        prices = new int[n.length + 1];
        System.arraycopy(n, 0, prices, 1, n.length);
    }

    public int bottomUp(int n) {
        int[] array = new int[n + 1];

        for (int i = 1; i <= n; i++) {
            array[i] = prices[i];
            for (int j = 1; j < i; j++) {
                array[i] = Math.min(array[i - j] + prices[j], array[i]);
            }
        }

        return array[n];
    }
}

File: src/main/java/baekjoon/dp/basic/카드 구매하기2.kt
package baekjoon.dp.basic

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader

/**
 * https://www.acmicpc.net/problem/16194
 */
fun main(args: Array<String>) {
    println(`카드 구매하기2`().solution(4, intArrayOf(1,5,6,7)) == 4)
    println(`카드 구매하기2`().solution(5, intArrayOf(10,9,8,7,6)) == 6)
    println(`카드 구매하기2`().solution(10, intArrayOf(1,1,2,3,5,8,13,21,34,55)) == 5)
    println(`카드 구매하기2`().solution(10, intArrayOf(5,10,11,12,13,30,35,40,45,47)) == 26)
    println(`카드 구매하기2`().solution(4, intArrayOf(5,2,8,10)) == 4)

    try {
        val fn = `카드 구매하기2`()
        val br = BufferedReader(InputStreamReader(System.`in`))
        val count = br.readLine().toInt()
        val prices = br.readLine().split(" ").stream()
            .mapToInt { it.toInt() }
            .toArray()
        println(fn.solution(count, prices))
    } catch (e: IOException) {/* */}
}

class `카드 구매하기2` {
    fun solution(n: Int, prices: IntArray): Int {
        val dp = IntArray(n + 1)

        for (i in 1..n) {
            dp[i] = prices[i - 1]
            for (j in 1..i) {
                dp[i] = dp[i].coerceAtMost(prices[j - 1] + dp[i - j])
            }
        }

        return dp[n]
    }
}